File: Basic operations on Linked List/C++/SinglyLinkedList.cpp
#include<bits/stdc++.h>
using namespace std;

struct node {
    int data;
    node * next;
};

class linked_list{
    private:
    node * head;
    node * tail;
    
    public:
    linked_list(){
        head = NULL;
        tail = NULL;
    }
    
    void createNode(int val);
    void displayList();
    void insert_start(int val);
    void insert_end(int val);
    void insert_at_pos(int p, int val);
    void delete_first();
    void delete_last();
    void reverse_list();
};

void linked_list::createNode(int value){
    node *temp = new node;
    temp->data = value;
    temp->next = NULL;
    
    if(head==NULL){
        head = temp;
        tail = temp;
        temp = NULL;
    }
    
    else {
        tail->next = temp;
        tail = temp;
    }
}

void linked_list::displayList(){
    node *temp ;//= new node;
    temp = head;
    while(temp != NULL){
        cout << temp->data<<"\t";
        temp = temp->next;
    }
    cout<<"\n";
}

void linked_list::insert_start(int value){
    node *temp = new node;
    temp->data = value;
    temp->next = head;
    head = temp;
}

void linked_list::insert_end(int value){
    node *temp = new node;
    temp->data = value;
    temp->next = NULL;
    tail->next = temp;
    tail = temp;
}

void linked_list::insert_at_pos(int pos, int value){
    node *temp = new node;
    node *cur = new node;
    node *pre = new node;
    cur = head;
    int count = 0;
    while(cur->next != NULL){
        cur = cur->next;
        count++;
    }
    if((count+1)>pos){
        cur = head;
        for(int i=1; i<pos; i++){
            pre = cur;
            cur = cur->next;
        }
        temp->data = value;
        temp->next = cur;
        pre->next = temp;
    }
    else{
        cout<<"Position should be less than or equal to "<<(count+1)<<"\n";
    }

}

void linked_list::delete_first(){
    node *temp;
    temp = head;
    head = head->next;
    delete temp;
}

void linked_list::delete_last(){
    node *cur;
    node *pre;
    cur = head;
    while(cur->next != NULL){
        pre = cur;
        cur = cur->next;
    }
    tail = pre;
    tail->next = NULL;
    delete cur;
}

void linked_list::reverse_list(){
    node *pre = new node;
    node *cur = new node;
    node *nxt = new node;
    cur = head;
    pre = NULL;
    nxt = NULL;
    
    while(cur != NULL){
        nxt = cur->next;
        cur->next = pre;
        pre = cur;
        cur = nxt;
    }
    head = pre;
}

int main(){
    int val;
    val = 10;
    linked_list ele;
    ele.createNode(val);
    ele.insert_end(11);
    ele.insert_start(9);
    ele.insert_at_pos(2,18);
    ele.displayList();
    ele.reverse_list();
    ele.displayList();
    return 0;
}

File: Basic operations on Linked List/Python/SinglyLinkedList.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None
        
class LinkedList:
    def __init__(self):
        self.head = None

    def printList(self):
        temp = self.head;
        while(temp):
            print(temp.data, end=" ")
            temp = temp.next
        print("\n")
            
    def insert_start(self, value):
        temp = Node(value)
        temp.next = self.head
        self.head = temp
        
    def insert_end(self, value):
        temp = Node(value)
        if self.head is None:   #If list is empty
            self.head = temp
        
        current = self.head
        while(current.next):    #Else iterate in list to the last
            current = current.next
        current.next = temp
        temp.next = None
        
    def insert_at_pos(self, pos, value):
        temp = Node(value)
        cur = self.head
        count = 0
        while(cur.next):
            count = count+1
            cur = cur.next
        if((count+1)>=pos and pos !=1 ):   
            cur = self.head
            for i in range (1,pos):
                pre = cur
                cur = cur.next
            pre.next = temp
            temp.next = cur
        
        elif(pos==1):
            self.insert_start(value)
        
        else:
            print("Position should be less than",count+2)
            
    def delete_first(self):
        temp = self.head
        self.head = temp.next
        
    def delete_last(self):
        cur = self.head
        while(cur.next):
            pre = cur
            cur = cur.next
        pre.next = None
        
    def delete_at_pos(self, pos):
        cur = self.head
        count = 0
        while(cur.next):
            count = count+1
            cur = cur.next
        
        if(pos==1):
            self.delete_first()
            
        elif((count+1)>=pos and pos !=1 ):   
            cur = self.head
            for i in range (1,pos):
                pre = cur
                cur = cur.next
            pre.next = cur.next
        
        else:
            print("Position should be less than ", count+2)
            
    def reverse_list(self):
        cur = self.head
        pre = None
        while(cur):
            nxt = cur.next
            cur.next = pre
            pre = cur
            cur = nxt
        self.head = pre
        
if __name__ == '__main__':
    
    element = LinkedList()
    element.head = Node(9)
    element.insert_start(8)
    element.insert_end(10)
    element.insert_at_pos(1,18)
    element.printList()
    element.reverse_list()
    element.printList()


